crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -153,9 +153,12 @@ impl<'interner> TypeChecker<'interner> {
                 let start_span = self.interner.expr_span(&for_expr.start_range);
                 let end_span = self.interner.expr_span(&for_expr.end_range);
 
-                let mut unify_loop_range = |actual_type, span| {
+                let mut unify_loop_range = |actual_type: &Type, span: Span| {
                     let expected_type = if self.is_unconstrained() {
-                        Type::FieldElement(CompTime::new(self.interner))
+                        Type::PolymorphicInteger(
+                            CompTime::new(self.interner),
+                            Shared::new(TypeBinding::Bound(actual_type.clone())),
```

Comment:
Creating a new polymorphic integer here for the loop start and end would allow them to be different types. Ie. `for i in 0u23 .. 7u64` would be allowed.

Also, if it is a design decision that we want to allow loops over integer variables, we should allow them in non-unconstrained functions as well. Assuming they are still comptime of course.